src/Domain/Amr/UpdateAmrDecision.cpp

```diff
@@ -0,0 +1,118 @@
+#include "Domain/Neighbors.hpp"  // IWYU pragma: keep
+#include "ErrorHandling/Assert.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+
+template <size_t VolumeDim>
+class OrientationMap;
+
+namespace amr {
+
+template <size_t VolumeDim>
+bool update_amr_decision(
+    gsl::not_null<std::array<amr::Flag, VolumeDim>*> my_current_amr_flags,
+    const Element<VolumeDim>& element, const ElementId<VolumeDim>& neighbor_id,
+    const std::array<amr::Flag, VolumeDim>& neighbor_amr_flags) noexcept {
+  const auto& element_id = element.id();
+  bool my_amr_decision_changed = false;
+  bool neighbor_found = false;
+  std::array<size_t, VolumeDim> my_desired_levels =
+      amr::desired_refinement_levels(element_id, *my_current_amr_flags);
+
+  for (const auto& direction_neighbors : element.neighbors()) {
+    const Neighbors<VolumeDim>& neighbors_in_dir = direction_neighbors.second;
+
+    if (1 == neighbors_in_dir.ids().count(neighbor_id)) {
+      neighbor_found = true;
+      const Direction<VolumeDim> direction_to_neighbor =
+          direction_neighbors.first;
+      const OrientationMap<VolumeDim>& orientation_of_neighbor =
+          neighbors_in_dir.orientation();
+      const std::array<size_t, VolumeDim> neighbor_desired_levels =
+          amr::desired_refinement_levels_of_neighbor(
+              neighbor_id, neighbor_amr_flags, orientation_of_neighbor);
+
+      // update flags if my element wants to be two or more levels
+      // coarser than the neighbor in any dimension
+      for (size_t d = 0; d < VolumeDim; ++d) {
+        if (amr::Flag::Split == gsl::at(*my_current_amr_flags, d) or
+            gsl::at(my_desired_levels, d) >=
+                gsl::at(neighbor_desired_levels, d)) {
+          continue;
+        }
+        const size_t difference =
+            gsl::at(neighbor_desired_levels, d) - gsl::at(my_desired_levels, d);
+        ASSERT(difference < 4 and difference > 0,
+               "neighbor level = " << gsl::at(neighbor_desired_levels, d)
+                                   << ", my level = "
+                                   << gsl::at(my_desired_levels, d));
+        if (amr::Flag::Join == gsl::at(*my_current_amr_flags, d)) {
+          if (3 == difference) {
+            // My split neighbor wants to split, so I need to split to keep
+            // refinement levels within one.
+            gsl::at(*my_current_amr_flags, d) = amr::Flag::Split;
+            gsl::at(my_desired_levels, d) += 2;
+            my_amr_decision_changed = true;
+          } else if (2 == difference) {
+            // My split neighbor wants to stay the same, or my neighbor
+            // split, so I need to stay the same to keep refinement levels
+            // within one.
+            gsl::at(*my_current_amr_flags, d) = amr::Flag::DoNothing;
+            ++gsl::at(my_desired_levels, d);
+            my_amr_decision_changed = true;
+          }
+        } else {  // I am not splitting or joining...
```

Comment:
sorry I don't understand what the confusion is about: the comment after the else, the contents of the else block, the whole function...?

Reply:
I was reading "I am not splitting or joining..." as "nothing in this code block will make me split or join" but in the if block it says "... so I need to split" which was confusing because it went against what I thought I understood earlier.

Reply:
ah okay; the comment is a relic from Bitbucket code where the proceeding if-clause was much longer and it helped remind what was the condition was to enter the else.  Since the if-clause is close, I have just removed the comment.